pkg/kube/krt/dynamicjoin.go

```diff
@@ -0,0 +1,436 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package krt
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/maps"
+	"istio.io/istio/pkg/ptr"
+	"istio.io/istio/pkg/util/sets"
+)
+
+// DynamicCollection is the interface for a collection that can have
+// other collections added or removed at runtime. Member collections
+// are keyed by the collection name.
+type DynamicCollection[T any] interface {
+	Collection[T]
+	AddOrUpdateCollection(c Collection[T])
+	RemoveCollection(c Collection[T])
```

Comment:
crazy or maybe not crazy idea: do we we want the interface to be imperative commands, or to have `NewJoinCollection(collections krt.Collection[Collection[T]])`?

Note that staticCollection has Update/Remove operations, so a user should be able to get the equivalent imperative workflow if they want.

I don't think we need to do this if it makes things more complicate though, but its plausible it makes things _less_ complex

Reply:
My only concern with the nested approach is that it now turns into a collection of a collection of T instead of a joined collection of T. I'm open to trying it out if you feel strongly though

Reply:
I meant that is how you construct it but it joins them for you

input col[col[T]]
output col[t]

Reply:
Ok I'm testing this locally and it's looking _really_ clean; great idea! My next commit will have both implementations and we can compare